### DISCUSSION TITLE
How to Simulate Electromagnetic Fields Generated by Point Sources in a Vacuum

### DISCUSSION THREAD

**Discussion by yiming1996** [NanoComp/meep#2306] [Q&A]
Hi,all
I use a frequency domain solver to simulate a point source in a vacuum. However, the results are quite different from those calculated by empirical formulas. So how to simulate electromagnetic fields generated by point sources in a vacuum?

Thanks!

The code is as follows:
`import meep as mp
import numpy as np
#import h5py
#import matplotlib.pyplot as plt
#resolution = 50 # 50 pixels per unit a.
resolution = 50
sxy = 20
dpml = 2

cell = mp.Vector3(sxy, sxy, 0) # just work in 2D for this
freq = 0.1
sources = [mp.Source(src=mp.ContinuousSource(freq),
                     center=mp.Vector3(x=0, y=0, z=0),
                     component=mp.Ez, 
                     amplitude=1.0) ] # 1mA source
geometry = [
        mp.Block(mp.Vector3(mp.inf, mp.inf, mp.inf), # define an infinite block
                     center=mp.Vector3(0, 0, 0), # centered at the origin
                     material=mp.Medium(epsilon=1))
] # that is a vacuum
pml_layers = [mp.PML(dpml)]
nonpml_vol = mp.Volume(mp.Vector3(), size=mp.Vector3(sxy - 2 * dpml, sxy - 2 * dpml, 0))
complex_fields = True
sim = mp.Simulation(cell_size=cell,
                    boundary_layers=pml_layers,
                    geometry=geometry,
                    sources=sources,
                    resolution=resolution,
                    force_complex_fields=complex_fields)
sim.init_sim()
sim.solve_cw(1e-10, 10000, 10)

ez_dat = sim.get_array(vol=nonpml_vol, component=mp.Ez)

ez_data_real = np.real(ez_dat)
ez_data_imag = np.imag(ez_dat)

np.savetxt('test5.txt',ez_data_real, fmt='%e', delimiter=',')
np.savetxt('test6.txt',ez_data_imag, fmt='%e', delimiter=',')`

The results are as follows:
[https://i.imgur.com/Zm8UWya.jpg](url)

**Comment by smartalecH** [comment]
Your link is broken. 

That is indeed a point source. But in 2D. So really it's an infinite line source extending in the z-direction. The analytic solution for this is much different than for a point source in 3d. 

Is there any reason you want to use the frequency-domain solver over the time-domain solver?

Note that the frequency-domain solver doesn't have any preconditioners... and the Maxwell problem is very ill-conditioned. As discussed in the meep paper, it may be worth it for highly resonant problems. But only if you need the full field profiles, as there are series extrapolation methods in the time domain that are probably even more efficient for extracting things like Q factor etc.
